# Task 3

First, update the following cell to have a link to your Homework 1 GitHub repository:


My Github Repo: 

https://github.com/hmchilders/eds220-hwk-1.git



In this task, make at least 5 commits at moments you think are adequate. We'll look at your repo to check commit history.

Add comments for all your code. Err on the side of commenting too much for now. 

Comments in each exercise and commits will count towards your grade.

This task covers topics we will see on Monday Oct. 9. The best strategy to solve it is to get together with your classmates and work on it together after Monday's lecture.


## 1
For this task we are going to use data about  [Western Indian Ocean Coral Diversity (McClanahan, 2023)](https://knb.ecoinformatics.org/view/doi%3A10.5063%2FF1K35S3H) from the Knowledge Network for Biocomplexity (KNB) data repository. 

a. Navigate to the dataset's link and explore the data package. 
 
b. Use the next markdown cell to add a brief description of the dataset, including a citation, date of access, and a link to the archive. 


This dataset looks at the current coral reef patterns can coral taxa in the Western Indian Ocean, and uses a machine learning program to predict the state of coral reefs in 2050. They did this by looking at the impact of climate change, water quality, direct human-resource extraction, and variable selection processes on numbers of coral taxa.

Tim McClanahan. (2023). Western Indian Ocean Coral Diversity. Knowledge Network for Biocomplexity. Date Accessed: 9 October 2023. https://knb.ecoinformatics.org/view/doi%3A10.5063%2FF1K35S3H

## 2
In the next code cell, import the pandas package using standard abbreviation in a code cell.

Read in the data from the URL using the `pd.read_csv` function and store it as `coral_div` like this:

```
# read in data
coral_div = pd.read_csv('the URL goes here')
```

In [1]:
#import pandas
import pandas as pd

# read in data
coral_div = pd.read_csv('https://knb.ecoinformatics.org/knb/d1/mn/v2/object/urn%3Auuid%3Ae7d12a5a-f1a9-4a3e-aaa5-5222f67e799d')

## 3

Use attributes and methods of the `coral_div` data frame to obtain the following information about it:

a. how many rows and columns does it have?

b. what are the column names?

c. what are the data types of the columns?

d. display the first 8 rows with all the column names. For this one, you will need to udpate a `pandas` option (see discussion section)

e. some inital info you are interested in. 

Use the code cells marked a-e. For each, add a comment explaining what is the ouput.

In [5]:
# a
coral_div.shape

(575, 42)

In [10]:
# b
coral_div.columns

Index(['PU_ID', 'PU_Lat', 'PU_Lon', 'Number_of_genera', 'Country', 'Ecoregion',
       'Management', 'Habitat', 'Observer', 'Depth', 'Coralcover',
       'SSTRateofRise', 'PARmax', 'Calcite', 'Dis_oxygen', 'mean.waveNRJ',
       'Salinity_mean', 'andrello_reef_value', 'andrello_nutrients',
       'mean.npp', 'TT_market_hrs', 'TT_pop_hrs', 'cumDHW', 'SST_Kurtosis',
       'SST_Median', 'ChlorA_median', 'andrello_sediments', 'Current_vel_mean',
       'Diff_attenuation', 'Grav_NP', 'Grav_NC', 'PH', 'SST_bimodality',
       'SST_Skewness', 'ClimateStressModel', 'Netflow', 'Indegree',
       'Outdegree', 'Retention', 'min_yr', 'max_yr', 'no_replicates'],
      dtype='object')

In [19]:
# c
coral_div.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 575 entries, 0 to 574
Data columns (total 42 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   PU_ID                575 non-null    int64  
 1   PU_Lat               575 non-null    float64
 2   PU_Lon               575 non-null    float64
 3   Number_of_genera     575 non-null    int64  
 4   Country              575 non-null    object 
 5   Ecoregion            575 non-null    object 
 6   Management           575 non-null    object 
 7   Habitat              530 non-null    object 
 8   Observer             575 non-null    object 
 9   Depth                575 non-null    float64
 10  Coralcover           575 non-null    int64  
 11  SSTRateofRise        575 non-null    float64
 12  PARmax               575 non-null    float64
 13  Calcite              575 non-null    float64
 14  Dis_oxygen           575 non-null    float64
 15  mean.waveNRJ         575 non-null    flo

In [15]:
# d
coral_div.head(n=8)

PU_ID     PU_Lat    PU_Lon  Number_of_genera   Country  \
0      7  39.076711 -5.490011                35  Tanzania   
1      7  39.076711 -5.490011                22  Tanzania   
2      7  39.076711 -5.490011                25  Tanzania   
3      7  39.076711 -5.490011                26  Tanzania   
4     10  39.054253 -5.512514                27  Tanzania   
5     10  39.054253 -5.512514                29  Tanzania   
6     15  39.054253 -5.535017                19  Tanzania   
7     39  39.009337 -5.647518                28  Tanzania   

                  Ecoregion   Management           Habitat Observer  Depth  \
0  East African Coral Coast  UnfishedLow             Crest      Tim   11.0   
1  East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira    2.0   
2  East African Coral Coast  UnfishedLow             Slope  Nyawira    4.0   
3  East African Coral Coast  UnfishedLow             Slope  Nyawira    3.0   
4  East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira    1.5   
5  East African Coral Coast  UnfishedLow  Lagoon/Back reef      Tim    1.5   
6  East African Coral Coast   Restricted             Crest      Tim    8.0   
7  East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira    3.0   

   ...  SST_bimodality  SST_Skewness  ClimateStressModel   Netflow  Indegree  \
0  ...        0.462556     -0.128136            0.596375  0.144498     98.25   
1  ...        0.462556     -0.128136            0.596375  0.144498     98.25   
2  ...        0.462556     -0.128136            0.596375  0.144498     98.25   
3  ...        0.462556     -0.128136            0.596375  0.144498     98.25   
4  ...        0.461297     -0.134548            0.629000  0.011721    112.75   
5  ...        0.461297     -0.134548            0.629000  0.011721    112.75   
6  ...        0.461297     -0.134548            0.595600  0.011721    112.75   
7  ...        0.464874     -0.158601            0.626000 -0.063087     96.50   

   Outdegree  Retention  min_yr  max_yr  no_replicates  
0      74.75      14.00    2019    2019              1  
1      74.75      14.00    2010    2015              2  
2      74.75      14.00    2015    2015              1  
3      74.75      14.00    2016    2019              2  
4      51.50     183.75    2010    2019              3  
5      51.50     183.75    2015    2015              1  
6      51.50     183.75    2019    2019              1  
7      40.75      35.75    2015    2019              3  

[8 rows x 42 columns]

In [17]:
# e
coral_div.info

<bound method DataFrame.info of      PU_ID     PU_Lat     PU_Lon  Number_of_genera       Country  \
0        7  39.076711  -5.490011                35      Tanzania   
1        7  39.076711  -5.490011                22      Tanzania   
2        7  39.076711  -5.490011                25      Tanzania   
3        7  39.076711  -5.490011                26      Tanzania   
4       10  39.054253  -5.512514                27      Tanzania   
..     ...        ...        ...               ...           ...   
570  11681  32.878336 -27.011417                19  South Africa   
571  11682  32.878336 -27.011417                18  South Africa   
572  11683  32.878336 -27.011417                19  South Africa   
573  11683  32.878336 -27.011417                19  South Africa   
574  11683  32.878336 -27.011417                21  South Africa   

                    Ecoregion   Management           Habitat Observer  Depth  \
0    East African Coral Coast  UnfishedLow             Crest      Tim  

## 4

Some of the columns have data type `int64` and some have data type `float64`. What is the difference between these `int64` and `float64`?


The main difference between the int64 and float64 data types is that int64 represents an interger and float64 represents a float. An integer is a number with no decimal points, a float is a number with decimal points.

## 5

The `Country` column indicates the country of reef location. Explain in your words what is the ouput of `coral_div.Country` and `coral_div.Country.unique()`. 

The coral_div.Country code will provide you with a list of all 575 entries for that columns, where as coral_div.Country.unique() will provide you with a list of each unique country. So the .unique() attribute wont print entries that have multiples.

## 6

Check the dataset's [metadata file README_md.doc](https://knb.ecoinformatics.org/view/doi:10.5063/F1K35S3H) to find which column has information about who made an observation. How many observers collected this data? HINT: use the `len` Python function. Not sure what `len` does? Use `?`.

In [21]:
# solve exercise here
len(coral_div['Observer'].unique())

16

## 7

The `Coralcover` column indicates the percent of hard coral cover in each 6km reef cell surveyed. What data is `coral_div[coral_div['Coralcover'] >= 50]` selecting?

This code only selects rows from the datasets where the percent of hard corl cover is at least 50%. To say that in another way, This code only selects rows from the dataset where the value in the 'Coralcover' column was greater than or equal to 50.

## 8

Use code to find if there are there any sites with a 30% coral cover in Tanzania.

In [30]:
# solve exercise here

coral_div[coral_div['Coralcover'] == 30]

PU_ID     PU_Lat     PU_Lon  Number_of_genera     Country  \
63     693  39.391121  -5.782490                10    Tanzania   
64     697  39.391121  -5.804983                22    Tanzania   
122    867  39.368664  -4.656797                23       Kenya   
123    867  39.368664  -4.656797                26       Kenya   
124    868  39.391121  -4.656797                24       Kenya   
125    868  39.391121  -4.656797                20       Kenya   
126    868  39.391121  -4.656797                22       Kenya   
127    872  39.436037  -4.679331                22       Kenya   
139   1186  55.246386 -21.012018                19     Reunion   
210   2377  49.968784 -15.717807                15  Madagascar   
341   4294  39.211458  -4.814515                37    Tanzania   
342   4299  39.233916  -4.837043                26    Tanzania   
343   4302  39.211458  -4.859570                31    Tanzania   
344   4303  39.301290  -4.859570                23    Tanzania   
377   4906  39.660616  -4.972195                33    Tanzania   
388   5049  39.615700  -5.490011                14    Tanzania   
389   5050  39.638158  -5.490011                18    Tanzania   

                           Ecoregion   Management           Habitat  Observer  \
63          East African Coral Coast       Fished  Lagoon/Back reef   Nyawira   
64          East African Coral Coast       Fished  Lagoon/Back reef   Nyawira   
122         East African Coral Coast  UnfishedLow  Lagoon/Back reef   Nyawira   
123         East African Coral Coast  UnfishedLow  Lagoon/Back reef       Tim   
124         East African Coral Coast  UnfishedLow  Lagoon/Back reef   Nyawira   
125         East African Coral Coast  UnfishedLow  Lagoon/Back reef   Nyawira   
126         East African Coral Coast  UnfishedLow  Lagoon/Back reef       Tim   
127         East African Coral Coast   Restricted             Slope       Tim   
139                Mascarene Islands   Restricted               NaN  Mireille   
210  Western and Northern Madagascar       Fished               NaN       Tim   
341         East African Coral Coast       Fished             Slope   Nyawira   
342         East African Coral Coast       Fished             Slope   Nyawira   
343         East African Coral Coast       Fished             Slope   Nyawira   
344         East African Coral Coast       Fished             Slope   Nyawira   
377         East African Coral Coast       Fished  Lagoon/Back reef   Nyawira   
388         East African Coral Coast       Fished  Lagoon/Back reef   Nyawira   
389         East African Coral Coast       Fished  Lagoon/Back reef   Nyawira   

         Depth  ...  SST_bimodality  SST_Skewness  ClimateStressModel  \
63    5.000000  ...        0.490512     -0.101724            0.655600   
64    9.166667  ...        0.490769     -0.100365            0.683050   
122   2.666667  ...        0.448886     -0.036279            0.727025   
123   1.000000  ...        0.448886     -0.036279            0.727025   
124   4.000000  ...        0.450158     -0.035417            0.723350   
125   3.187500  ...        0.450158     -0.035417            0.723350   
126   4.000000  ...        0.450158     -0.035417            0.723350   
127  29.000000  ...        0.451827     -0.035988            0.628700   
139   2.500000  ...        0.568952     -0.009900            0.212200   
210   8.500000  ...        0.596164     -0.153239            0.470300   
341  11.000000  ...        0.451293     -0.048992            0.608900   
342  12.500000  ...        0.451293     -0.048992            0.601725   
343   7.000000  ...        0.452402     -0.050477            0.595200   
344  11.000000  ...        0.453806     -0.048007            0.602300   
377   9.000000  ...        0.488396     -0.074229            0.671900   
388   4.000000  ...        0.489921     -0.109523            0.657000   
389   6.000000  ...        0.489921     -0.109523            0.618825   

      Netflow  Indegree  Outdegree  Retent

## 9

We are interested in looking at the PH of sea water (`PH` column),  mean salinity (PSS) (`Salinity_mean` column), and depth of survey location in meters (`Depth` column). Select this data. 

In [9]:
# solve exercise here

## 10
The `max_yr` column indicates the year of last sampling at a given site. Select the PH, salinity, and depth for sites where sampling ended in 2017. 


In [10]:
# solve exercise here

## 11
Use `iloc` to select rows 95 throuhg 100 (including 100) in the 17th, 20th, and 28th columns and assign these values to a new variable called `subset`. HINT: the ouput columns should be `Salinity_mean`, `mean.npp` and `Current_vel_mean`. Are you indexing from 0?

In [11]:
# solve exercise here

## 12 

Run `subset.isna()`. Explain what this code does and what the output represents.

*Update this markdown cell with your answer.*

## 13
Run `subset.isna().any()`. Explain what this code does and what the output represents. HINT: It could be helpful to check the [documentation of `pandas.Sries.any`](https://pandas.pydata.org/docs/reference/api/pandas.Series.any.html).


*Update this markdown cell with your answer.*

## 14 

Select the rows in `subset` that have NA in the `Current_vel_mean` column. HINT: you can use your answer for 9.

In [12]:
# solve exercise here